### PR TITLE
Reclaim User Map Example

### PR DESCRIPTION
**Pull request by ash-burnt**
## Contract Details
Please append all the required information below for the contract(s) being added to `contracts.json`.

Required fields and their descriptions:
- `name`: Contract name (required)
- `description`: Brief description of the contract's purpose
- `code_id`: Contract code ID on mainnet
- `hash`: Contract hash in UPPERCASE
- `release`:
  - `url`: URL to the release/commit (e.g., https://github.com/org/repo/releases/tag/v1.0.0)
  - `version`: Version tag or first 7 chars of commit hash
- `author`:
  - `name`: Organization name
  - `url`: Organization website URL
- `governance`: "Genesis" or proposal number
- `deprecated`: true if contract is deprecated (mixed inline with active contracts)

Example JSON structure:
```json
{
  "name": "",
  "description": "",
  "code_id": "",
  "hash": "",
  "release": {
    "url": "",
    "version": ""
  },
  "author": {
    "name": "",
    "url": ""
  },
  "governance": "",
  "deprecated": false
}
```

### Finding Code ID and Hash
To find the latest code ID and hash:
1. Run the verification tool which will show all code IDs on chain:
   ```bash
   node scripts/verify-code-ids.js
   ```
2. The new code ID will be shown in the mismatches as "exists on chain but not in contracts.json"
3. You can also query the code hash via the chain's RPC endpoint:
   ```bash
   xiond query wasm code-info <code-id> --node https://rpc.xion-mainnet-1.burnt.com
   ```

### Documentation Updates
The README.md is automatically generated from `contracts.json`. After making changes:

1. Ensure you have the required dependencies:
   - Node.js: https://nodejs.org/
   - jq: `brew install jq` (macOS) or `apt-get install jq` (Ubuntu/Debian)

2. Run the convert script to validate and update the README:
   ```bash
   ./convert.sh
   ```

3. Commit both the `contracts.json` and generated `README.md` changes

⚠️ Important Notes:
- Do not edit README.md manually. All changes must be made through `contracts.json`
- Pull requests with manual README edits will be automatically rejected by CI
- If you forget to run `./convert.sh` locally, the CI will fail with a "README out of sync" error

### Validation
The `convert.sh` script automatically performs these validations:
- All required fields are present and properly formatted
- Hash is 64 characters and uppercase hex
- URLs are valid HTTPS links
- Code IDs are unique
- Contracts are ordered by code_id (both active and deprecated contracts follow the same ordering)
- README.md stays in sync with contracts.json

If any validation fails, the script will show specific error messages to help you fix the issues.

### Checklist
- [ ] Added entry to `contracts.json` with all required fields
- [ ] Contract name is clear and descriptive
- [ ] Description explains the contract's purpose
- [ ] Code ID matches the mainnet deployed code
- [ ] Hash is in uppercase and matches the stored code
- [ ] Release URL points to the correct tag/commit
- [ ] Version matches the release tag or commit hash
- [ ] Author information is correct with valid URL
- [ ] Governance field correctly references proposal or "Genesis"
- [ ] Deprecated flag is set appropriately
- [ ] Entry is placed in code_id order (regardless of deprecated status)
- [ ] Ran `./convert.sh` and fixed any validation errors
- [ ] Both `contracts.json` and generated `README.md` are included in the commit

## Summary by CodeRabbit

* **New Features**
  * Added a contract that records verified claim values for wallet addresses.
  * Added proof-based updates using a configurable verification service and claim key.
  * Added queries to retrieve an individual value, list registered addresses, or view the complete address-to-value map.
  * Added contract configuration and migration message support.
* **Documentation**
  * Added schema generation for contract messages and responses.

<h3>Greptile Summary</h3>

This change adds a Reclaim-backed user map contract. Two release and query paths need attention: the new package is absent from the lockfile, preventing locked artifact builds, and the collection queries return every stored user or map entry without a bounded page.

<h3>Confidence Score: 3/5</h3>

The contract cannot be released through the locked build path, and unbounded collection responses will grow with every historical user.

Two independent non-security failures remain: reproducible locked-build failure and unbounded collection retrieval.

**Files Needing Attention:** Cargo.lock, contracts/reclaim_user_map/Cargo.toml, contracts/reclaim_user_map/src/contract.rs, contracts/reclaim_user_map/src/msg.rs

<details><summary><h3><a href="https://www.greptile.com/trex">T-Rex</a> T-Rex Logs</h3></summary>

**What T-Rex did**
- T-Rex produced a proof for a posted P1 finding and linked the related Cargo validation outputs as artifacts.
- T-Rex produced proofs for two additional P1 findings.
- The saved focused check read the current query-message definitions and query implementation and passed after confirming GetUsers and GetMap have no start\_after or limit fields, and that both paths collect and serialize all entries, while a native Cargo test attempt could not fetch the pinned private Git dependency.
- T-Rex executed the claim-update rollback verification script and captured before/after logs, noting a verifier failure that caused the transaction to roll back.

<a href="https://app.greptile.com/trex/runs/20048542/artifacts"><picture><source media="(prefers-color-scheme: dark)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/ViewAllArtifactsDark.svg?v=4"><source media="(prefers-color-scheme: light)" srcset="https://greptile-static-assets.s3.amazonaws.com/badges/ViewAllArtifacts.svg?v=4">View all artifacts</picture></a>

<sub><a href="https://www.greptile.com/trex">T-Rex</a> Ran code and verified through T-Rex</sub>
</details>

<sub>Reviews (2): Last reviewed commit: ["Merge branch &#39;main&#39; into example/reclaim..."](https://github.com/burnt-labs/contracts/commit/f873bd15dd9132936b367c88fce461b3d61844eb) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=55568748)</sub>

> Greptile also left **3 inline comments** on this PR.

### CODE REVIEW

**Comment by crucible-burnt** [comment]
## 🔍 Crucible Security Review

### Summary
Adds Reclaim User Map example contract for integrating with the Reclaim protocol. Stores user data based on verified Reclaim proofs.

### Security Assessment
- **Risk Level**: Low (example contract for protocol integration)
- **Instantiation validation**: Address validation and non-empty claim key check ✓
- **JSON parsing**: Safe with error handling for missing extractedParameters and keys
- **State management**: Simple mapping of user address to verified value
- **External calls**: Delegates to verification contract after state update (proper ordering)
- **No re-entrancy risk**: State updated before external call

### Immunefi Pattern Check
- ✅ Proper error handling
- ✅ No unvalidated external calls
- ✅ No state mutation vulnerabilities
- ✅ Claim key must match extracted parameter

### False Report Risk
- Mark as example/educational contract to prevent incorrect security assumptions
- Add comments explaining Reclaim protocol flow

### Code Quality
- Clean match statements
- Proper error types for each failure mode
- Good state separation

### Recommendation
**Approve** - Clean integration example with proper error handling. Add documentation clarifying this is an example contract.

**Review by crucible-burnt — COMMENTED** [review]
## 🔍 Crucible Security Review

### Summary
New contract implementing a user→value map populated via Reclaim protocol proof verification. Users submit proofs that are verified by an external Reclaim contract, and extracted claim values are stored per-user.

### Security Assessment
- **Risk Level**: Medium

**Findings:**

1. **⚠️ State stored before proof verification** (`src/contract.rs:51-55`): `USER_MAP.save()` is called before the Reclaim verification sub-message executes. However, this is **safe** because CosmWasm executes atomically — if the Reclaim contract rejects the proof, the entire transaction reverts including the `USER_MAP.save()`. Not a vulnerability, but worth a comment for clarity.

2. **JSON parsing uses `HashMap<&str, Value>`** (`src/contract.rs:44`): The `context` field is parsed with borrowed string keys from the `value` input. This is fine for lifetime management since `value` lives for the duration of the function.

3. **`value.to_string()` on `serde_json::Value`** (`src/contract.rs:50`): This will include JSON quotes for string values (e.g., `"\"hello\""` instead of `"hello"`). If consuming contracts/queries expect unquoted strings, this could cause issues. Consider using `v.as_str().unwrap_or(&v.to_string())` for string values.

4. **No access control on `ExecuteMsg::Update`**: Any address can call `Update` and store a value for themselves. This is by design (self-sovereign claims), but verify this is intentional — there's no admin override or pause mechanism.

5. **No value size limit**: The `verified_value` string stored per user has no size bound. A malicious Reclaim proof with an extremely large extracted parameter could cause storage bloat. Consider adding a max value length check.

6. **Pinned `reclaim_xion` dependency** (`Cargo.toml:21`): Uses a specific git rev `5c67c33d` — good for reproducibility, but should be periodically updated.

7. **`always_fail` getrandom override** (`src/lib.rs:13-17`): Standard CosmWasm pattern, correctly implemented.

### Immunefi Pattern Check
- Trust boundary is the external Reclaim verification contract — if that contract has vulnerabilities, this contract's stored values could be based on invalid proofs
- No direct cryptographic operations — delegates to Reclaim contract
- No fee/gas manipulation vectors

### False Report Risk
- The "state before verification" pattern could invite incorrect bug reports about TOCTOU or state manipulation. Consider adding a comment explaining CosmWasm's atomic execution guarantees.

### Code Quality Notes
- Error types are minimal but appropriate
- No unit tests included — recommend adding tests for JSON parsing edge cases, empty context, missing keys
- Consider adding an admin/pause mechanism for emergency response

### Status
Functional implementation with correct security model. Main recommendations: document the atomic execution assumption, fix `to_string()` quoting behavior, add value size limits, and include unit tests.

**Review by crucible-burnt — COMMENTED** [review]
## 🔍 Crucible Security Review

Reclaim User Map example contract (188 lines). **Risk Level: Low**. No security concerns.

File: contracts/reclaim_user_map/src/contract.rs [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+    deps: DepsMut,
+    _env: Env,
+    info: MessageInfo,
+    msg: InstantiateMsg,
+) -> Result<Response, ContractError> {
+    deps.api.addr_validate(msg.verification_addr.as_str())?;
+    VERIFICATION_ADDR.save(deps.storage, &msg.verification_addr)?;
+    if msg.claim_key.is_empty() {
+        return Err(ClaimKeyInvalid)
+    }
```
[the remark this answers is not in the document]
Comment (crucible-burnt) [reply contracts/reclaim_user_map/src/contract.rs:25]: Checked the Rust syntax directly: `return Err(ClaimKeyInvalid)` is valid as the tail expression of this `if` block; no semicolon is required. This is not a compile error, so I am leaving the code unchanged.

File: contracts/reclaim_user_map/src/contract.rs [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+pub fn execute(
+    deps: DepsMut,
+    _: Env,
+    info: MessageInfo,
+    msg: ExecuteMsg,
+) -> ContractResult<Response> {
+    match msg {
+        ExecuteMsg::Update { value } => {
+            // validate JSON
+            let context: HashMap<&str, Value> = serde_json::from_str(&value.proof.claimInfo.context)?;
```
[the remark this answers is not in the document]
Comment (crucible-burnt) [reply contracts/reclaim_user_map/src/contract.rs]: Checked the lifetime against the actual input ownership: `serde_json` can deserialize object keys into `&str` borrowed from the input string, and that input remains alive for the full use of this map. An owned `String` could be a robustness preference, but the current type is not inherently invalid or a compile failure, so no corrective change is warranted.
